### ISSUE TITLE
Lowercased FTP commands not properly identified

### ISSUE THREAD

**Issue by lordhoto** [proftpd/proftpd#1023]
### What I Did

I did a setup of a very simple ProFTPd Instance (config attached) on SLES12. I used telnet to issue lowercase commands to the FTP server to rename a file:
```
$ telnet localhost 21
Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 ProFTPD Server (ProFTPd-Case-Test) [::1]
user test
331 Password required for test
pass test
230 User test logged in
rnfr test-orig
350 File or directory exists, ready for destination name
rnto test-new
501 Bad sequence of commands
quit
221 Goodbye.
Connection closed by foreign host.
```

### What I Expected/Wanted

I expected ProFTPd to handle lowercase commands the same as uppercase commands. This is according to the [FTP RFC](https://tools.ietf.org/html/rfc959) as far as I understand:
```
[...] Upper and lower case alphabetic characters are to be treated identically. [...]
```

I need this feature because some clients issue lowercase commands and I can't make users switch their FTP clients.

### ProFTPD Version and Configuration

Output of proftpd -V:
```
Compile-time Settings:
  Version: 1.3.6b (maint)
  Platform: LINUX [Linux 4.12.14-122.20-default x86_64]
  Built With:
    configure  '--host=x86_64-suse-linux-gnu' '--build=x86_64-suse-linux-gnu' '--program-prefix=' '--prefix=/usr' '--exec-prefix=/usr' '--bindir=/usr/bin' '--sbindir=/usr/sbin' '--sysconfdir=/etc' '--datadir=/usr/share' '--includedir=/usr/include' '--libdir=/usr/lib64' '--libexecdir=/usr/lib' '--localstatedir=/var' '--sharedstatedir=/usr/com' '--mandir=/usr/share/man' '--infodir=/usr/share/info' '--disable-dependency-tracking' '--disable-static' '--bindir=/usr/sbin' '--libexecdir=/usr/lib64/proftpd' '--sysconfdir=/etc/proftpd' '--localstatedir=/run/proftpd' '--enable-sendfile' '--enable-ctrls' '--enable-dso' '--enable-facl' '--enable-ipv6' '--enable-nls' '--enable-openssl' '--with-lastlog' '--with-includes=/usr/include/mysql:/usr/include/pgsql' '--with-shared=mod_auth_otp:mod_ban:mod_copy:mod_ctrls_admin:mod_deflate:mod_digest:mod_dnsbl:mod_dynmasq:mod_exec:mod_geoip:mod_ifsession:mod_ifversion:mod_ldap:mod_load:mod_log_forensic:mod_qos:mod_quotatab:mod_quotatab_file:mod_quotatab_ldap:mod_quotatab_radius:mod_quotatab_sql:mod_radius:mod_ratio:mod_readme:mod_rewrite:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_site_misc:mod_snmp:mod_sql:mod_sql_mysql:mod_sql_odbc:mod_sql_passwd:mod_sql_postgres:mod_sql_sqlite:mod_statcache:mod_tls:mod_tls_fscache:mod_tls_memcache:mod_tls_redis:mod_tls_shmcache:mod_unique_id:mod_wrap2:mod_wrap2_file:mod_wrap2_redis:mod_wrap2_sql' '--disable-ident' '--disable-strip' 'build_alias=x86_64-suse-linux-gnu' 'host_alias=x86_64-suse-linux-gnu' 'CFLAGS=-fmessage-length=0 -grecord-gcc-switches -fstack-protector -O2 -Wall -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -g -D_GNU_SOURCE -DLDAP_DEPRECATED' 'CXXFLAGS=-fmessage-length=0 -grecord-gcc-switches -fstack-protector -O2 -Wall -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -g -D_GNU_SOURCE -DLDAP_DEPRECATED'

  CFLAGS: -g2 -fmessage-length=0 -grecord-gcc-switches -fstack-protector -O2 -Wall -D_FORTIFY_SOURCE=2 -funwind-tables -fasynchronous-unwind-tables -g -D_GNU_SOURCE -DLDAP_DEPRECATED -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib  -rdynamic  -L/usr/lib64 -L/usr/lib/postgresql10/lib64
  LIBS: -lacl  -lssl -lcrypto -L$(top_srcdir)/lib/libcap -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl 

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd/proftpd.pid
    Scoreboard File:
      /run/proftpd/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib64/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    - curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    - Memcache support
    + ncursesw support
    + NLS support
    - Redis support
    - Sodium support
    + OpenSSL support (FIPS enabled)
    - PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10

```

ProFTPd Configuration:
```
ServerName ProFTPd-Case-Test
ServerType standalone

DefaultServer on

User ftp
Group ftp

PidFile        /tmp/proftpd/proftpd.pid
ScoreboardFile /tmp/proftpd/proftpd.scoreboard
DelayTable     /tmp/proftpd/proftpd.delay

<Global>
	AuthPAM off
</Global>

AuthUserFile  /tmp/proftpd/proftpd.passwd
AuthGroupFile /tmp/proftpd/proftpd.group
AuthOrder mod_auth_file.c

RequireValidShell no

DefaultRoot /tmp/proftpd/proftpd_root/%u

DirFakeUser  on
DirFakeGroup on
```

Passwd file:
```
test:$1$qI5AFb9B$Wd/bEtPqFl3IAB6P/5mkb/:40:49::/tmp/proftpd/proftpd_root/test:/bin/false
```

Group file:
```
ftp:x:49:test
```

In /tmp/proftpd_root/test I create a file for the client to rename. It's called "test-orig".

### Open points

Is there any configuration parameters I need to set to make lowercase rnto work? Did I do something wrong? Thank you in advance.

Oddly enough with uppercase RNTO but all other commands lowercase it works:
```
telnet localhost 21
Trying ::1...
Connected to localhost.
Escape character is '^]'.
220 ProFTPD Server (ProFTPd-Case-Test) [::1]
user test
331 Password required for test
pass test
230 User test logged in
rnfr test-orig
350 File or directory exists, ready for destination name
RNTO test-new
250 Rename successful
quit
221 Goodbye.
Connection closed by foreign host.
```

**Comment by lordhoto** [comment]
I did further analysis on the issue. I upgrade to 1.3.7c, but it didn't help. With gdb and browsing the sources a bit I found out the following.

When an RNTO (all uppercase) command is issued by the client the "cmd" Parameter for "core_pre_any" has the "cmd_id" 25 (which seems to be PR_CMD_RNTO_ID, i.e. what I would expect):
```
Breakpoint 1, core_pre_any (cmd=0x2415ee8) at mod_core.c:3481
3481    MODRET core_pre_any(cmd_rec *cmd) {
(gdb) p *cmd
$11 = {pool = 0x2415ea0, server = 0x23e3c68, config = 0x0, tmp_pool = 0x242f6a0, argc = 2, arg = 0x2415f88 "test-new", argv = 0x2415fa8, group = 0x0, cmd_class = 48,
  stash_index = -1, stash_hash = 0, notes = 0x2413cd8, cmd_id = 25, is_ftp = 1, protocol = 0x4ba740 "FTP"}
```

When an rnto (all lowercase) command is issued by the client, the "cmd" Parameter has a "cmd_id" of -1 (invalid?):
```
Breakpoint 1, core_pre_any (cmd=0x2415ee8) at mod_core.c:3481
3481    MODRET core_pre_any(cmd_rec *cmd) {
(gdb) p *cmd
$10 = {pool = 0x2415ea0, server = 0x23e3c68, config = 0x0, tmp_pool = 0x2430190, argc = 2, arg = 0x2415f88 "test-new", argv = 0x2415fa8, group = 0x0, cmd_class = 48,
  stash_index = -1, stash_hash = 0, notes = 0x242f6e8, cmd_id = -1, is_ftp = 1, protocol = 0x4ba740 "FTP"}
```
Setting the "cmd_id" to "25" in gdb allows the command to be executed successfully.

I would naively assume that the cmd_id for lowercase commands is not properly set. Maybe it is done before the commands have been converted to uppercase? I don't know. I hope this information helps.

**Comment by Castaglia** [comment]
[This PR](https://github.com/proftpd/proftpd/pull/1024) should fix things.

The bug here is not specific to `RNTO`.  When ProFTPD reads in text, it converts that text to a `cmd_rec`.  Part of parsing that text is identifying (and caching!) the received command, giving it a command ID (`cmd->cmd_id`).  The very first time the received text gets a command ID assigned is [here](https://github.com/proftpd/proftpd/blob/master/src/cmd.c#L464), in `pr_cmd_is_http()`, where we check for protocol mismatches.  The text has *not* been uppercased here, so no valid command ID is assigned; the "invalid" command ID value of -1 is cached.  Hence the failure to recognize that `rnto` is a valid command.

The fix, then, is to ensure that the first word of the command text has been uppercased _just before_ performing our protocol mismatch checks.  That way, the correct command ID is assigned and cached.

**Comment by lordhoto** [comment]
Yeah, this also explains another oddity I saw. In the logs the PASS command is logged with password when it's issued lowercase. Thanks for the fix.

Will be another fix release for 1.3.6? Or is that only for security updates? Thank you.

**Comment by Castaglia** [comment]
There will be another maintenance release of 1.3.6, yes -- and I'll backport this fix to the 1.3.6 branch.

**Comment by Castaglia** [comment]
Fixed in master, and backported to 1.3.6.  Thanks!

**Comment by lordhoto** [comment]
Thank you for fixing it so quickly.
